fx.go

```diff
@@ -18,4 +18,150 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-package fx // import "go.uber.org/fx"
+package fx
+
+import (
+	"io"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"github.com/uber-go/tally"
+
+	"go.uber.org/dig"
+	"go.uber.org/fx/config"
+	"go.uber.org/fx/metrics"
+	"go.uber.org/fx/ulog"
+	"go.uber.org/zap"
+)
+
+// Component is a function that initializes and returns objects to be shared across the code
+type Component interface{}
+
+// Module is a building block or primary function performed by a service built on fx
+type Module interface {
+	Name() string
+	Constructor() []Component
+	Stop()
+}
+
+// Service controls the lifecycle of an fx service
+type Service struct {
+	g              *dig.Graph
+	modules        []Module
+	components     []Component
+	scopeCloser    io.Closer
+	logger         *zap.Logger
+	loggerCloserFn func()
+}
+
+// scopeInit is used by the metrics package
+// TODO: Remove after metrics refactor
+type scopeInit struct {
+	config config.Provider
+}
+
+// Name returns the name
+func (c *scopeInit) Name() string {
+	return c.config.Get(config.ServiceNameKey).AsString()
+}
+
+// Config returns the config
+func (c *scopeInit) Config() config.Provider {
+	return c.config
+}
+
+// New creates a service with the provided modules
+func New(modules ...Module) *Service {
+	s := &Service{
+		g:       dig.New(),
+		modules: modules,
+	}
+
+	// load config for module creation and include it in the graph
+	cfg := config.DefaultLoader.Load()
+	// TODO: need to pull latest dig for direct Interface injection fix
+	s.g.MustRegister(func() config.Provider { return cfg })
```

Comment:
fx is using an older version of dig that doesnt work with interfaces very well. See https://github.com/uber-go/dig/issues/27. There is a TODO already to change this once latest version of dig is pulled in.